### PR TITLE
CORDA-3600 Add flowIO request to checkpoint

### PR DESCRIPTION
**Pull request by williamvigorr3**
FlowMonitor updates the checkpoint database with stuck flow information instead of simply logging.

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt [hunk 1]
```diff
@@ -53,9 +57,16 @@ internal class FlowMonitor(
...
-            flow.ioRequest()?.let { request -> logger.info(warningMessageForFlowWaitingOnIo(request, flow, suspensionDuration)) }
+    @VisibleForTesting
+    fun updateDbAndLogWaitingFlows() {
+        database.transaction {
+            for ((flow, suspensionDuration) in waitingFlowDurations(suspensionLoggingThreshold)) {
+                val ioRequest = flow.ioRequest()
+                ioRequest?.let { request -> logger.info(warningMessageForFlowWaitingOnIo(request, flow, suspensionDuration)) }
+                if (ioRequest != null) {
+                    checkpointStorage.updateFlowIoRequest(flow.id, ioRequest)
+                }
```
Comment (lankydan) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt]: This can go into the `let` block above:

```kotlin
flow.ioRequest().ioRequest?.let {
    logger.info(warningMessageForFlowWaitingOnIo(it, flow, suspensionDuration)) }
    checkpointStorage.updateFlowIoRequest(flow.id, it)
}
```

File: node/src/test/kotlin/net/corda/node/services/persistence/DBCheckpointStorageTests.kt [hunk 2]
```diff
@@ -347,6 +350,27 @@ class DBCheckpointStorageTests {
         }
     }
 
+    @Test(timeout = 300_000)
+    fun `Check that flow io request can be added to a checkpoint`() {
```
Comment (lankydan) [inline node/src/test/kotlin/net/corda/node/services/persistence/DBCheckpointStorageTests.kt]: Rename the test to:

`checkpoint can be updated with flow io request information`

File: node/src/test/kotlin/net/corda/node/services/persistence/DBCheckpointStorageTests.kt [hunk 3]
```diff
@@ -347,6 +350,27 @@ class DBCheckpointStorageTests {
...
+        val (id, checkpoint) = newCheckpoint(1)
+        database.transaction {
+            val serializedFlowState = checkpoint.flowState.checkpointSerialize(context = CheckpointSerializationDefaults.CHECKPOINT_CONTEXT)
+            checkpointStorage.addCheckpoint(id, checkpoint, serializedFlowState)
+            val checkpoint = checkpointStorage.getDBCheckpoint(id)
+            assertNull(checkpoint?.ioRequestType)
```
Comment (lankydan) [inline node/src/test/kotlin/net/corda/node/services/persistence/DBCheckpointStorageTests.kt]: Do !! to make sure it is not the checkpoint that is null instead of the `ioRequestType`

File: node/src/test/kotlin/net/corda/node/services/persistence/DBCheckpointStorageTests.kt [hunk 4]
```diff
@@ -347,6 +350,27 @@ class DBCheckpointStorageTests {
...
+        }
+        val instantNow = Instant.now()
+        database.transaction {
+            checkpointStorage.updateFlowIoRequest(id, FlowIORequest.Sleep(instantNow))
+        }
+        database.transaction {
+            val checkpoint = checkpointStorage.getDBCheckpoint(id)
+            assertNotNull(checkpoint?.ioRequestType)
+            val classFromCheckpoint = checkpoint?.ioRequestType!!
+            assert(classFromCheckpoint == FlowIORequest.Sleep::class.java)
```
Comment (lankydan) [inline node/src/test/kotlin/net/corda/node/services/persistence/DBCheckpointStorageTests.kt]: Can we use `assertEquals(FlowIORequest.Sleep::class.java, classFromCheckpoint)`

File: node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt [hunk 5]
```diff
@@ -69,6 +73,17 @@ class FlowFrameworkTests {
     private lateinit var notaryIdentity: Party
     private val receivedSessionMessages = ArrayList<SessionTransfer>()
 
+    private fun dbCheckpointStorageWithStubPerfRecorder() : DBCheckpointStorage {
```
Comment (lankydan) [inline node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt]: Just call it `checkpointStorage` no one will care about the performance recorder.

File: node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt [hunk 6]
```diff
@@ -166,46 +181,83 @@ class FlowFrameworkTests {
             val terminationSignal = Semaphore(0)
```
Comment (lankydan) [inline node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt]: Can we give the test a better name.

File: node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt [hunk 7]
```diff
@@ -166,46 +181,83 @@ class FlowFrameworkTests {
...
-            assertEquals(1, aliceFlowMonitor.waitingFlowDurations(Duration.ZERO).toSet().size)
-            assertEquals(0, bobFlowMonitor.waitingFlowDurations(Duration.ZERO).toSet().size)
+            val aliceFlowId = aliceFlow.id
+            val bobFlowId = bobNode.smm.allStateMachines[0].stateMachine.id
+
+            aliceFlowMonitor.updateDbAndLogWaitingFlows()
+            bobFlowMonitor.updateDbAndLogWaitingFlows()
+
+            aliceNode.database.transaction {
+                assert(dbCheckpointStorageWithStubPerfRecorder().getCheckpointIoRequest(aliceFlowId) == FlowIORequest.Receive::class.java)
```
Comment (lankydan) [inline node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt]: Use `assertEquals`

File: node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt (hunk 5 above)
Comment (lankydan) [reply node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt]: And store it in a variable

File: node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt [hunk 8]
```diff
@@ -166,46 +181,83 @@ class FlowFrameworkTests {
...
+            }
+            bobNode.database.transaction {
+                assert(dbCheckpointStorageWithStubPerfRecorder().getCheckpointIoRequest(bobFlowId) == FlowIORequest.Receive::class.java)
+            }
         }
     }
 
     @Test(timeout=300_000)
 	fun `FlowMonitor flow is running`() { // flow is running a "take a long time" task
```
Comment (lankydan) [inline node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt]: Can we give this a better name as well please.

File: node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt [hunk 9]
```diff
@@ -227,6 +228,17 @@ class DBCheckpointStorage(private val checkpointPerformanceRecorder: CheckpointP
...
     }
 
+    override fun updateFlowIoRequest(id: StateMachineRunId, ioRequest: FlowIORequest<*>) {
+        val checkpoint = getDBCheckpoint(id)
+        checkpoint?.ioRequestType = ioRequest.javaClass
+        currentDBSession().update(checkpoint)
+    }
+
+    @VisibleForTesting
+    fun getDBCheckpoint(id: StateMachineRunId): DBFlowCheckpoint? {
```
Comment (lankydan) [inline node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt]: I personally would rather we just didn't have this function, but I am aware that makes some duplicated code in testing classes but then it isnt too bad.

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt [hunk 10]
```diff
@@ -53,9 +57,16 @@ internal class FlowMonitor(
         }
     }
 
-    private fun logFlowsWaitingForParty() {
-        for ((flow, suspensionDuration) in waitingFlowDurations(suspensionLoggingThreshold)) {
-            flow.ioRequest()?.let { request -> logger.info(warningMessageForFlowWaitingOnIo(request, flow, suspensionDuration)) }
+    @VisibleForTesting
+    fun updateDbAndLogWaitingFlows() {
+        database.transaction {
```
Comment (lankydan) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt]: I think it makes sense to use a single database transaction (like you have done) 👍

**Review by lankydan — COMMENTED** [review]
Looks about right. Incoming tidy up comments!

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt [hunk 11]
```diff
@@ -53,9 +57,16 @@ internal class FlowMonitor(
...
+            for ((flow, suspensionDuration) in waitingFlowDurations(suspensionLoggingThreshold)) {
+                val ioRequest = flow.ioRequest()
+                ioRequest?.let { request -> logger.info(warningMessageForFlowWaitingOnIo(request, flow, suspensionDuration)) }
+                if (ioRequest != null) {
```
Comment (lankydan) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt]: We should probably also retrieve checkpoint to test if the `ioRequest` has changed (is `null` or something else`). That way we can remove an extra update. It should be retrieving it from the hibernate cache so performance _should_ be alright. That being said, I am not sure if this will improve the performance or not... (I don't know hibernate performance well enough).

File: node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt (hunk 9 above)
Comment (mnesbit) [reply node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt]: What tests actually need this?

Comment (williamvigorr3) [reply node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt]: Some of the `FlowFrameworkTests` that use the FlowMonitor and a test I added to`DBCheckpointStorageTests`.  Actually this would be easy to remove (this used to be needed before @lankydan made some changes to #600 which I overlooked when rebasing).

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt (hunk 1 above)
Comment (williamvigorr3) [reply node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt]: Done.

File: node/src/test/kotlin/net/corda/node/services/persistence/DBCheckpointStorageTests.kt (hunk 2 above)
Comment (williamvigorr3) [reply node/src/test/kotlin/net/corda/node/services/persistence/DBCheckpointStorageTests.kt]: Done.

File: node/src/test/kotlin/net/corda/node/services/persistence/DBCheckpointStorageTests.kt (hunk 3 above)
Comment (williamvigorr3) [reply node/src/test/kotlin/net/corda/node/services/persistence/DBCheckpointStorageTests.kt]: Oops done.

File: node/src/test/kotlin/net/corda/node/services/persistence/DBCheckpointStorageTests.kt (hunk 4 above)
Comment (williamvigorr3) [reply node/src/test/kotlin/net/corda/node/services/persistence/DBCheckpointStorageTests.kt]: Done.

File: node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt (hunk 5 above)
Comment (williamvigorr3) [reply node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt]: Done that.

File: node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt (hunk 6 above)
Comment (williamvigorr3) [reply node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt]: Done.

File: node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt (hunk 7 above)
Comment (williamvigorr3) [reply node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt]: Done.

File: node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt (hunk 8 above)
Comment (williamvigorr3) [reply node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt]: Done this..

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt (hunk 11 above)
Comment (williamvigorr3) [reply node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt]: Okay I have done this in DBCheckpointStorage (as we were already extracting and updating there).

File: node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt [hunk 12]
```diff
@@ -250,6 +257,10 @@ class DBCheckpointStorage(private val checkpointPerformanceRecorder: CheckpointP
         }
     }
 
+    private fun getDBCheckpoint(id: StateMachineRunId): DBFlowCheckpoint? {
```
Comment (lankydan) [inline node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:251]: I think this should use `find` instead of `get`. You can then use it in some of the other functions.

File: node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt [hunk 13]
```diff
@@ -161,51 +176,87 @@ class FlowFrameworkTests {
...
+            val bobFlowId = bobNode.smm.allStateMachines[0].stateMachine.id
+
+            aliceFlowMonitor.updateDbAndLogWaitingFlows()
+            bobFlowMonitor.updateDbAndLogWaitingFlows()
+            // both flows are suspended on a receive from the counter party
+            aliceNode.database.transaction {
+                assert(dbCheckpointStorage.getCheckpointIoRequest(aliceFlowId) == FlowIORequest.Receive::class.java)
+            }
+            bobNode.database.transaction {
+                assert(dbCheckpointStorage.getCheckpointIoRequest(bobFlowId) == FlowIORequest.Receive::class.java)
```
Comment (lankydan) [inline node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt]: `assertEquals`

File: node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt [hunk 14]
```diff
@@ -228,8 +228,10 @@ class DBCheckpointStorage(private val checkpointPerformanceRecorder: CheckpointP
 
     override fun updateFlowIoRequest(id: StateMachineRunId, ioRequest: FlowIORequest<*>) {
         val checkpoint = getDBCheckpoint(id)
-        checkpoint?.ioRequestType = ioRequest.javaClass
-        currentDBSession().update(checkpoint)
+        if (checkpoint?.ioRequestType != ioRequest) {
```
Comment (kyriathar) [inline node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt]: Really minor thing but would it maybe be better to write this `if` block as `checkpoint?.let` { ...

File: node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt (hunk 13 above)
Comment (williamvigorr3) [reply node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt]: Oops sorry I thought I fixed all of those. Done now.

File: node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt (hunk 14 above)
Comment (williamvigorr3) [reply node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt]: Yeah I have done this. I think I understood what you meant but if there is a cleaner way let me know.

File: node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt (hunk 12 above)
Comment (williamvigorr3) [reply node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:251]: This should be done now.

File: node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt (hunk 14 above)
Comment (kyriathar) [reply node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt]: Yes you are right in that its not very clear, if we could please replace `currentDBSession().update(checkpoint)` with `currentDBSession().update(it)  `. 
To make it a bit more clear, maybe, it could be written like: `it.apply {ioRequestType = ioRequest.javaClass}` i think.

File: node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt [hunk 15]
```diff
@@ -227,6 +226,16 @@ class DBCheckpointStorage(private val checkpointPerformanceRecorder: CheckpointP
         currentDBSession().update(updateDBCheckpoint(id, checkpoint, serializedFlowState))
     }
 
+    override fun updateFlowIoRequest(id: StateMachineRunId, ioRequest: FlowIORequest<*>) {
+        val checkpoint = getDBCheckpoint(id)
+        checkpoint?.let {
```
Comment (lankydan) [inline node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt]: Makes sense that this also does:

```kotlin
val entity = requireNotNull(getDBCheckpoint(id)) {
            "Checkpoint with id: ${id.uuid} does not exist in database."
        }
```

Like you did in `updateDBCheckpoint`. The current null check can then be removed.

Comment (williamvigorr3) [reply node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt]: Done.

**Comment by mnesbit** [comment]
I am a bit confused by this. FlowIO requests are captured in the stack of the fiber. I can understand needing to put some tracking strings in the DB, but do we really want to serialize/deserialize objects to the DB just for informational purposes? Definitely anything we write to DB is trumped by the info in the fiber.

**Comment by lankydan** [comment]
We are writing the flow io request to the database to provide something that users can query for. I agree that the information in the fiber is the _correct_ information but that requires any filtering based on the flow io request to be done in memory.

Not doing this work in-memory is why we have done it in the `FlowMonitor`. Only filling in the flow io request field of flows that have been suspended for longer than a period of time.

We shouldn't be doing a serialization/deserialization here. We are determining whether we need to update the flow io request using in-memory data and then updating only that column in the table.

If we want to switch to doing this all in memory (which will make it 100% accurate) we need to decide sooner rather than later due to the effects querying the data.

**Review by lankydan — DISMISSED** [review]

**Comment by rick-r3** [comment]
I thought we were going to flip the flow IO column to be a string and have the class short name ?

**Comment by lankydan** [comment]
@rick-r3 I talked to @roastario about this, and from his perspective, it does not matter whether it is a class name or a string. We have stuck with the `Class` since it makes the code tidier. We can always switch over to a string later on if stefano's implementation is hindered by doing so.

**Comment by mnesbit** [comment]
I am worried that Class might not be reloadable after upgrades, so we might lose the info for the incompatible flows.

**Comment by lankydan** [comment]
Sounds like we should switch to a string then ^^

**Comment by williamvigorr3** [comment]
Okay I have switched to a String using simpleName.

File: node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt [hunk 16]
```diff
@@ -336,7 +345,7 @@ class DBCheckpointStorage(private val checkpointPerformanceRecorder: CheckpointP
             this.status = checkpoint.status
             this.compatible = checkpoint.compatible
             this.progressStep = checkpoint.progressStep
-            this.ioRequestType = checkpoint.flowIoRequest
+            this.ioRequestType = checkpoint.flowIoRequest?.simpleName?.take(MAX_SIMPLE_NAME_LENGTH)
```
Comment (lankydan) [inline node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt]: I think because we are in control of the names of the `FlowIORequest` classes that we don't need to worry about the length being hit. All of the current instances of `FlowIORequest` should be under the limit. If we manage to have a super long length name in the future we can change this code then (but i doubt we would manage to do that).

File: node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt [hunk 17]
```diff
@@ -382,7 +391,7 @@ class DBCheckpointStorage(private val checkpointPerformanceRecorder: CheckpointP
             result = result?.let { SerializedBytes<Any>(it.value) },
             status = status,
             progressStep = progressStep,
-            flowIoRequest = ioRequestType,
+            flowIoRequest = null,
```
Comment (lankydan) [inline node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt]: Lets just make `flowIoRequest` type a `String` and set it as such so we don't have problems loading from the database.

File: node/src/main/kotlin/net/corda/node/services/api/CheckpointStorage.kt [hunk 18]
```diff
@@ -41,4 +42,10 @@ interface CheckpointStorage {
...
      */
     fun getAllCheckpoints(): Stream<Pair<StateMachineRunId, Checkpoint.Serialized>>
-}
\ No newline at end of file
+
+    /**
+     * Get the simpleName of the flowIoRequest which the flow suspended on. Returns null if the flow didn't suspend
+     * on an flowIoRequest.
+     */
+    fun getFlowIoRequest(id: StateMachineRunId): String?
```
Comment (lankydan) [inline node/src/main/kotlin/net/corda/node/services/api/CheckpointStorage.kt]: We can remove this once `Checkpoint.flowIORequest` has become a string

**Review by lankydan — APPROVED** [review]
I think we have made it!

**Comment by williamvigorr3** [comment]
:champagne:
